tsdb/wlog/checkpoint.go

```diff
@@ -64,26 +64,50 @@ func LastCheckpoint(dir string) (string, int, error) {
 	return filepath.Join(dir, checkpoint.name), checkpoint.index, nil
 }
 
-// DeleteCheckpoints deletes all checkpoints in a directory below a given index.
+// CheckpointPrefix is the prefix used for checkpoint files.
+const CheckpointPrefix = "checkpoint."
+
+// CheckpointTempFileSuffix is the suffix used when creating temporary checkpoint files.
+const CheckpointTempFileSuffix = ".tmp"
+
+// DeleteCheckpoints deletes all checkpoints in a directory below a given index and any temporary checkpoints that might have
+// failed to be created.
 func DeleteCheckpoints(dir string, maxIndex int) error {
-	checkpoints, err := listCheckpoints(dir)
+	checkpoints, err := listCheckpoints(dir, true)
 	if err != nil {
 		return err
 	}
 
+	if len(checkpoints) == 0 {
+		return nil
+	}
+
+	// We can't guarantee that Checkpoint() will not be called at the same time as DeleteCheckpoints() so we only delete
+	// temporary checkpoints that are older than the most recent successful checkpoint.
+	var lastSuccessfulCheckpointIndex int
+	for lastSuccessfulCheckpointIndex = len(checkpoints) - 1; lastSuccessfulCheckpointIndex >= 0; lastSuccessfulCheckpointIndex-- {
+		if !strings.HasSuffix(checkpoints[lastSuccessfulCheckpointIndex].name, CheckpointTempFileSuffix) {
+			break
+		}
+	}
+	// We didn't find a successful checkpoint, keep the most recent temporary checkpoint.
+	if lastSuccessfulCheckpointIndex == -1 {
+		lastSuccessfulCheckpointIndex = checkpoints[len(checkpoints)-1].index
```

Comment:
`lastSuccessfulCheckpointIndex` is between 0 and `len(checkpoints)-1` if this `if` is not taken, but `.index` is taken from the path name on disk, so in that case the number could be much higher. 
And this number is compared against `.index` on line 103.
Is this correct?